### PR TITLE
Large federated shares

### PR DESCRIPTION
**Pull request by butonic**
Prevent timeouts when adding large federated shares:
* completely removes getShareinfo() and instead uses PROPFIND
* removes the Scanner, as it is no longer needed
* shareinfo.php is left in because it may be used by existing instances
  * a deprecation message is added to the JSON response
  - [ ] should we add sth like will be removed in OC 11?
* removes external.php because it predates the OCS api, but we might want to only log deprecation warnings, as I did for shareinfo.php

backport needed to 9.1

### CODE REVIEW

File: apps/files_sharing/ajax/shareinfo.php [hunk 1]
```diff
@@ -25,8 +25,17 @@
  *
  */
 
+/**
+ * @deprecated fetches all metadata, kept for compatibility with older instances
+ */
+
 OCP\JSON::checkAppEnabled('files_sharing');
 
+OC::$server->getLogger()->warning(
```
Comment (PVince81) [inline apps/files_sharing/ajax/shareinfo.php]: Not sure if we can afford this... likely to flood logs. Switch to debug maybe ?

Also an admin might not know what hundreds of different remotes are connecting to this instance as this is controlled by the users.

**Comment by PVince81** [comment]
> should we add sth like will be removed in OC 11?

not really as this was never officially public API...

Code changed look good, needs good testing, especially the "detect temporarily unavailable remote server" cases.

- [ ] TEST: accepting share without notification app enabled (just in case because of external.php)
- [ ] TEST: accepting share with notification app enabled (just in case because of external.php)
- [ ] TEST: accessing/syncing federated share
- [ ] TEST: sharer removes share access, StorageInvalidException, receiver's mount disappears at first access
- [ ] TEST: sharer changes password, StorageInvalidException, receiver's mount disappears at first access
- [ ] TEST: sharer OC in maintenance mode, StorageNotAvailableException, receiver's mount stays and returns 503. When problem resolved, share is accessible again.
- [ ] TEST: sharer OC returns 404 (move away all code), StorageNotAvailableException, receiver's mount stays and returns 503. When problem resolved, share is accessible again.

**Comment by PVince81** [comment]
possible extension: make the DAV storage read oc:size from PROPFIND and return it (if the scanner does support reading this value), something for another PR

**Comment by PVince81** [comment]
@butonic sadly it looks like we'll have to keep "ajax/external.php", it's used when mounting link shares... but will need to replace its "scanAll" call with something that only scans one level / the root entry.

**Comment by PVince81** [comment]
local testing went fine except for the public link shares

**Comment by PVince81** [comment]
@tomneedham can you help finishing this ?

**Comment by PVince81** [comment]
Pushed changes to bring back ajax/external.php but without using the removed `scanAll`.

Had a quick look at using `oc:size` but it will be a little bit more work: need the DAV class to override `Common::getMetaData()` because that method always uses -1 for folders.

**Comment by PVince81** [comment]
This is ready for re-review, manual tests passed.

I hereby approve of @butonic's commits, given that they require my additions.

@butonic please review my additional commits.

File: apps/files_sharing/ajax/shareinfo.php (hunk 1 above)
Comment (butonic) [inline apps/files_sharing/ajax/shareinfo.php]: My intention is to annoy admins to get them to upgrade older versions. And if they have to annoy another admin I am fine with it. Saves everyone problems and time in the long run.

Comment (PVince81) [inline apps/files_sharing/ajax/shareinfo.php]: following MS' footsteps with Win10 upgrades ? not sure...

**Comment by butonic** [comment]
@PVince81 changes make sense

https://github.com/owncloud/core/blob/f10d105ea2f2d0822693313ddcd6f9b2eaac6ebe/apps/files_sharing/appinfo/routes.php#L64-L65 is the place where we use ajax/external.php and make it available as /external

https://github.com/owncloud/core/blob/master/apps/files_sharing/js/external.js#L103 is the place where we POST to the url

**Comment by PVince81** [comment]
@butonic is that a thumbs up for merging or do you think we need further changes ?

**Comment by tomneedham** [comment]
Tested with two instances running this. Remote sharing a folder worked, accessed the files, and uploaded between the two instances.  👍

**Comment by PVince81** [comment]
let me rebase this for CI...

Backport to stable10 only or further ?

**Comment by butonic** [comment]
@PVince81 changes make sense

https://github.com/owncloud/core/blob/f10d105ea2f2d0822693313ddcd6f9b2eaac6ebe/apps/files_sharing/appinfo/routes.php#L64-L65 is the place where we use ajax/external.php and make it available as /external

https://github.com/owncloud/core/blob/master/apps/files_sharing/js/external.js#L103 is the place where we POST to the url

**Comment by PVince81** [comment]
@jfd let's do the route change separately as it's not related to the performance issue we want to solve and backport

**Comment by PVince81** [comment]
stable10: https://github.com/owncloud/core/pull/28604

@butonic should we backport further ?

**Comment by butonic** [comment]
I prefer getting 10.0.3 ready over a backport to 9.1. @felixboehm ?

File: apps/files_sharing/ajax/shareinfo.php (hunk 1 above)
Comment (bcutter) [reply apps/files_sharing/ajax/shareinfo.php]: I strongly vote for **reverting** those 4 lines (34-37). It´s annoying and - guess what - really helps NOTHING, especially when all the remote servers are hosted instances without admin access or even when they run Nextcloud.

**Pull request merged**